HOPE:If resolver1 return SERVFAIL, ROUTEDNS should treat as request failed and switch to resolver2.

[groups.group1]
  resolvers=["resolver1","resolver2"]
  type="fail-rotate"

NOW: If resolver1 return SERVFAIL, group1 will return SERVFAIL to downstream.

HOPE: If resolver1 return SERVFAIL, ROUTEDNS should treat as request failed and switch to resolver2. Then send the same requestion to resolver2.If both request failed, return SERVFAIL to downstream.

REASON: If some public DNS wait timeout, it will return SERVFAIL until get answer. Sometimes It is just network problem between NS server and public DNS server.

panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x18 pc=0x15bc0bd]

goroutine 40 [running]:
github.com/folbricht/routedns.(*FailRotate).Resolve(0xc0000725a0, 0xc000140360, 0xc0003bc10c, 0x10, 0x10, 0xc000159750, 0x12a0c6c, 0xc00003b860)
        /go/pkg/mod/github.com/folbricht/routedns@v0.0.0-20201107185713-f6f7f2700c3f/failrotate.go:44 +0x63d
github.com/folbricht/routedns.(*ResponseBlocklistIP).Resolve(0xc0001a6bd0, 0xc000140360, 0xc0003bc10c, 0x10, 0x10, 0xc0000a38c8, 0xc0000a38c8, 0xc0001a14ba)
        /go/pkg/mod/github.com/folbricht/routedns@v0.0.0-20201107185713-f6f7f2700c3f/response-blocklist-ip.go:60 +0x67
github.com/folbricht/routedns.(*ResponseBlocklistIP).Resolve(0xc0001a6c40, 0xc000140360, 0xc0003bc10c, 0x10, 0x10, 0x1, 0xc000111880, 0xc000028c00)
        /go/pkg/mod/github.com/folbricht/routedns@v0.0.0-20201107185713-f6f7f2700c3f/response-blocklist-ip.go:60 +0x67
github.com/folbricht/routedns.(*Router).Resolve(0xc000394870, 0xc000140360, 0xc0003bc10c, 0x10, 0x10, 0x10, 0x1, 0x2030000)
        /go/pkg/mod/github.com/folbricht/routedns@v0.0.0-20201107185713-f6f7f2700c3f/router.go:61 +0x370
github.com/folbricht/routedns.listenHandler.func1(0x1885220, 0xc0000fe8a0, 0xc000140360)
        /go/pkg/mod/github.com/folbricht/routedns@v0.0.0-20201107185713-f6f7f2700c3f/dnslistener.go:71 +0x6d3
github.com/miekg/dns.HandlerFunc.ServeDNS(0xc0001a7ce0, 0x1885220, 0xc0000fe8a0, 0xc000140360)
        /go/pkg/mod/github.com/miekg/dns@v1.1.32/server.go:37 +0x4b
github.com/miekg/dns.(*Server).serveDNS(0xc0001fd0e0, 0xc0000c0640, 0x3c, 0x3c, 0xc0000fe8a0)
        /go/pkg/mod/github.com/miekg/dns@v1.1.32/server.go:617 +0x2f7
github.com/miekg/dns.(*Server).serveTCPConn(0xc0001fd0e0, 0xc00038be40, 0x1885400, 0xc0000065d0)
        /go/pkg/mod/github.com/miekg/dns@v1.1.32/server.go:525 +0x219
created by github.com/miekg/dns.(*Server).serveTCP
        /go/pkg/mod/github.com/miekg/dns@v1.1.32/server.go:442 +0x365

You can not merge this branch, because it is not stable. It will crash after get SERVFAIL response. (My golang version is 1.15.3.)

--- a/doc/configuration.md
+++ b/doc/configuration.md
@@ -365,7 +365,7 @@ type = "round-robin"
 
 ### Fail-Rotate group
 
-In a Fail-Rotate group, one of the upstream resolvers or modifiers is active and receives all queries. If the active resolver fails, the next becomes active and the request is retried. If the last resolver fails the first becomes the active again. There's no time-based automatic fail-back.
+In a Fail-Rotate group, one of the upstream resolvers or modifiers is active and receives all queries. If the active resolver fails, i.e. no response or returns SERVFAIL, the next becomes active and the request is retried. If the last resolver fails the first becomes the active again. There's no time-based automatic fail-back.
 
 #### Configuration
 
@@ -385,7 +385,7 @@ type = "fail-rotate"
 
 ### Fail-Back group
 
-Similar to [fail-rotate](#Fail-Rotate-group) but will attempt to fall back to the original order (prioritizing the first) if there are no failures for a minute.
+Similar to [fail-rotate](#Fail-Rotate-group) but will attempt to fall back to the original order (prioritizing the first) if there are no failures for a minute. Failure means either no response or it returns SERVFAIL.
 
 #### Configuration
 

--- a/failback.go
+++ b/failback.go
@@ -77,7 +77,7 @@ func (r *FailBack) Resolve(q *dns.Msg, ci ClientInfo) (*dns.Msg, error) {
 		log.WithField("resolver", resolver.String()).Debug("forwarding query to resolver")
 		r.metrics.route.Add(resolver.String(), 1)
 		a, err := resolver.Resolve(q, ci)
-		if err == nil && a.Rcode != dns.RcodeServerFailure { // Return immediately if successful
+		if err == nil && (a == nil || a.Rcode != dns.RcodeServerFailure) { // Return immediately if successful
 			return a, err
 		}
 		log.WithField("resolver", resolver.String()).WithError(err).Debug("resolver returned failure")

--- a/failback.go
+++ b/failback.go
@@ -77,7 +77,7 @@ func (r *FailBack) Resolve(q *dns.Msg, ci ClientInfo) (*dns.Msg, error) {
 		log.WithField("resolver", resolver.String()).Debug("forwarding query to resolver")
 		r.metrics.route.Add(resolver.String(), 1)
 		a, err := resolver.Resolve(q, ci)
-		if err == nil { // Return immediately if successful
+		if err == nil && a.Rcode != dns.RcodeServerFailure { // Return immediately if successful
 			return a, err
 		}
 		log.WithField("resolver", resolver.String()).WithError(err).Debug("resolver returned failure")

--- a/failback_test.go
+++ b/failback_test.go
@@ -47,3 +47,24 @@ func TestFailBack(t *testing.T) {
 	require.Equal(t, 5, r1.HitCount())
 	require.Equal(t, 1, r2.HitCount())
 }
+
+func TestFailBackSERVFAIL(t *testing.T) {
+	// Build 2 resolvers that count the number of invocations
+	var ci ClientInfo
+	opt := StaticResolverOptions{
+		RCode: dns.RcodeServerFailure,
+	}
+	r1, err := NewStaticResolver("test-static", opt)
+	require.NoError(t, err)
+
+	r2 := new(TestResolver)
+
+	g := NewFailBack("test-fb", FailBackOptions{ResetAfter: time.Second}, r1, r2)
+	q := new(dns.Msg)
+	q.SetQuestion("test.com.", dns.TypeA)
+
+	// Send the first query, the first resolver will return SERVFAIL and the request will go to the 2nd
+	_, err = g.Resolve(q, ci)
+	require.NoError(t, err)
+	require.Equal(t, 1, r2.HitCount())
+}

--- a/failback_test.go
+++ b/failback_test.go
@@ -68,3 +68,18 @@ func TestFailBackSERVFAIL(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, 1, r2.HitCount())
 }
+
+func TestFailBackDrop(t *testing.T) {
+	var ci ClientInfo
+	r1 := NewDropResolver("test-drop")
+	r2 := new(TestResolver)
+
+	g := NewFailBack("test-fb", FailBackOptions{ResetAfter: time.Second}, r1, r2)
+	q := new(dns.Msg)
+	q.SetQuestion("test.com.", dns.TypeA)
+
+	// The query should be dropped, so no failover
+	_, err := g.Resolve(q, ci)
+	require.NoError(t, err)
+	require.Equal(t, 0, r2.HitCount())
+}

--- a/failrotate.go
+++ b/failrotate.go
@@ -41,7 +41,7 @@ func (r *FailRotate) Resolve(q *dns.Msg, ci ClientInfo) (*dns.Msg, error) {
 		log.WithField("resolver", resolver.String()).Trace("forwarding query to resolver")
 		r.metrics.route.Add(resolver.String(), 1)
 		a, err := resolver.Resolve(q, ci)
-		if err == nil && a.Rcode != dns.RcodeServerFailure { // Return immediately if successful
+		if err == nil && (a == nil || a.Rcode != dns.RcodeServerFailure) { // Return immediately if successful
 			return a, err
 		}
 		log.WithField("resolver", resolver.String()).WithError(err).Debug("resolver returned failure")

--- a/failrotate.go
+++ b/failrotate.go
@@ -41,7 +41,7 @@ func (r *FailRotate) Resolve(q *dns.Msg, ci ClientInfo) (*dns.Msg, error) {
 		log.WithField("resolver", resolver.String()).Trace("forwarding query to resolver")
 		r.metrics.route.Add(resolver.String(), 1)
 		a, err := resolver.Resolve(q, ci)
-		if err == nil { // Return immediately if successful
+		if err == nil && a.Rcode != dns.RcodeServerFailure { // Return immediately if successful
 			return a, err
 		}
 		log.WithField("resolver", resolver.String()).WithError(err).Debug("resolver returned failure")

--- a/failrotate_test.go
+++ b/failrotate_test.go
@@ -82,3 +82,18 @@ func TestFailRotateSERVFAIL(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, 1, r2.HitCount())
 }
+
+func TestFailRotateDrop(t *testing.T) {
+	var ci ClientInfo
+	r1 := NewDropResolver("test-drop")
+	r2 := new(TestResolver)
+
+	g := NewFailRotate("test-rotate", r1, r2)
+	q := new(dns.Msg)
+	q.SetQuestion("test.com.", dns.TypeA)
+
+	// The query should be dropped, so no failover
+	_, err := g.Resolve(q, ci)
+	require.NoError(t, err)
+	require.Equal(t, 0, r2.HitCount())
+}

--- a/failrotate_test.go
+++ b/failrotate_test.go
@@ -61,3 +61,24 @@ func TestFailRotate(t *testing.T) {
 	require.Equal(t, 5, r1.HitCount())
 	require.Equal(t, 5, r2.HitCount())
 }
+
+func TestFailRotateSERVFAIL(t *testing.T) {
+	// Build 2 resolvers that count the number of invocations
+	var ci ClientInfo
+	opt := StaticResolverOptions{
+		RCode: dns.RcodeServerFailure,
+	}
+	r1, err := NewStaticResolver("test-static", opt)
+	require.NoError(t, err)
+
+	r2 := new(TestResolver)
+
+	g := NewFailRotate("test-rotate", r1, r2)
+	q := new(dns.Msg)
+	q.SetQuestion("test.com.", dns.TypeA)
+
+	// Send the first query, the first resolver will return SERVFAIL and the request will go to the 2nd
+	_, err = g.Resolve(q, ci)
+	require.NoError(t, err)
+	require.Equal(t, 1, r2.HitCount())
+}

